File: Microsoft.Toolkit.Uwp.Input.GazeInteraction/ExpandCollapsePatternGazeTargetItem.cpp
#include "pch.h"
#include "ExpandCollapsePatternGazeTargetItem.h"
#include "ExpandCollapsePatternGazeTargetItem.g.cpp"
#include "winrt/Microsoft.UI.Xaml.Automation.h"
#include "winrt/Microsoft.UI.Xaml.Automation.Peers.h"
#include "winrt/Microsoft.UI.Xaml.Automation.Provider.h"

using namespace winrt;
using namespace winrt::Microsoft::UI::Xaml::Automation;
using namespace winrt::Microsoft::UI::Xaml::Automation::Peers;
using namespace winrt::Microsoft::UI::Xaml::Automation::Provider;

namespace winrt::Microsoft::Toolkit::Uwp::Input::GazeInteraction::implementation
{
    void ExpandCollapsePatternGazeTargetItem::Invoke()
    {
        auto peer = FrameworkElementAutomationPeer::FromElement(TargetElement());
        auto provider = peer.GetPattern(PatternInterface::ExpandCollapse).try_as<IExpandCollapseProvider>();
        switch (provider.ExpandCollapseState())
        {
        case ExpandCollapseState::Collapsed:
            provider.Expand();
            break;

        case ExpandCollapseState::Expanded:
            provider.Collapse();
            break;
        }
    }
}

File: Microsoft.Toolkit.Uwp.Input.GazeInteraction/ExpandCollapsePatternGazeTargetItem.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using Windows.UI.Xaml;
using Windows.UI.Xaml.Automation;
using Windows.UI.Xaml.Automation.Peers;
using Windows.UI.Xaml.Automation.Provider;

namespace Microsoft.Toolkit.Uwp.Input.GazeInteraction
{
    internal class ExpandCollapsePatternGazeTargetItem : GazeTargetItem
    {
        internal ExpandCollapsePatternGazeTargetItem(UIElement element)
            : base(element)
        {
        }

        internal override void Invoke()
        {
            var peer = FrameworkElementAutomationPeer.FromElement(TargetElement);
            var provider = peer.GetPattern(PatternInterface.ExpandCollapse) as IExpandCollapseProvider;
            switch (provider.ExpandCollapseState)
            {
                case ExpandCollapseState.Collapsed:
                    provider.Expand();
                    break;

                case ExpandCollapseState.Expanded:
                    provider.Collapse();
                    break;
            }
        }
    }
}